### PR TITLE
[WMS][12.0]Add stock_location_tray

### PR DESCRIPTION
**Pull request by guewen**
Extracted from stock_vertical_lift
(https://github.com/OCA/stock-logistics-warehouse/pull/633)

Add tray types to stock locations, automatically generates
sub-locations. Present them nicely with a custom widget.

### CODE REVIEW

File: stock_location_tray/models/stock_location.py [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+            tray = self
+        location = self.search(
+            [
+                ("id", "child_of", tray.ids),
+                # we receive positions counting from 0 but they are stored
+                # in the "human" format starting from 1
+                ("posx", "=", coordX + 1),
+                ("posy", "=", coordY + 1),
+            ]
+        )
```
Comment (sebalix) [inline stock_location_tray/models/stock_location.py:83]: Should we add a `location.ensure_one()` to enforce the expected result?

File: stock_location_tray/models/stock_location.py [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+                trays_to_update = location.tray_type_id.id != new_tray_type_id
+            # short-circuit this check if we already know that we have to
+            # update trays
+            if not trays_to_update and "cell_name_format" in vals:
+                new_format = vals.get("cell_name_format")
+                trays_to_update = location.cell_name_format != new_format
+            super(StockLocation, location).write(vals)
+            if trays_to_update:
+                self._update_tray_sublocations()
+            elif "posz" in vals and location.tray_type_id == "tray":
```
Comment (sebalix) [inline stock_location_tray/models/stock_location.py]: `location.tray_type_id == "tray"` something is wrong here I suppose?

File: stock_location_tray/models/stock_location.py [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+    def _tray_cell_matrix(self):
+        assert self.tray_type_id or self.cell_in_tray_type_id
+        if self.tray_type_id:
+            location = self
+        else:  # cell
+            location = self.location_id
+        cells = location.tray_type_id._generate_cells_matrix()
+        for cell in location.child_ids:
+            if cell.tray_cell_contains_stock:
+                cells[cell.posy - 1][cell.posx - 1] = 1
```
Comment (sebalix) [inline stock_location_tray/models/stock_location.py:196]:
```suggestion
                # 1 means used
                cells[cell.posy - 1][cell.posx - 1] = 1
```

File: stock_location_tray/views/stock_location_views.xml [hunk 4]
```diff
@@ -0,0 +1,48 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+
+    <record id="view_location_form_tray_type" model="ir.ui.view">
```
Comment (sebalix) [inline stock_location_tray/views/stock_location_views.xml]: Wee should keep the original XML-ID if possible (`view_location_form`), the namespace being the module now.

File: stock_location_tray/views/stock_location_views.xml [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+            <field name="posy" position="attributes">
+                <attribute name="attrs">{'readonly': [('cell_in_tray_type_id', '!=', False)]}</attribute>
+            </field>
+            <field name="posz" position="attributes">
+                <attribute name="attrs">{'readonly': [('cell_in_tray_type_id', '!=', False)]}</attribute>
+            </field>
+        </field>
+    </record>
+
+    <record id="view_location_search_vertical_lift" model="ir.ui.view">
```
Comment (sebalix) [inline stock_location_tray/views/stock_location_views.xml]: same

**Review by sebalix — COMMENTED** [review]
Code review, small comments

File: stock_location_tray/views/stock_location_views.xml (hunk 4 above)
Comment (guewen) [reply stock_location_tray/views/stock_location_views.xml]: I used to do that, but it's everywhere different, so I do not know anymore what's better :)
I'll use the original xmlid.

File: stock_location_tray/views/stock_location_views.xml (hunk 5 above)
Comment (guewen) [reply stock_location_tray/views/stock_location_views.xml]: plus remove vertical.lift from the name

**Comment by guewen** [comment]
@sebalix thanks for your review! pushed the changes

**Comment by guewen** [comment]
Merged in 13.0, may have been some changes there. If needed in 12.0, we should backport the 13.0 one.

**Pull request closed without merging**